src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -0,0 +1,2293 @@
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::declare_function
+
+  Inputs:
+    function_name - a name
+    type - a type
+    symbol_table - symbol table
+
+ Purpose: declare a function with the given name and type
+
+\*******************************************************************/
+
+void java_string_library_preprocesst::declare_function(
+  irep_idt function_name, const typet &type, symbol_tablet &symbol_table)
+{
+  auxiliary_symbolt func_symbol;
+  func_symbol.base_name=function_name;
+  func_symbol.is_static_lifetime=false;
+  func_symbol.mode=ID_java;
+  func_symbol.name=function_name;
+  func_symbol.type=type;
+  symbol_table.add(func_symbol);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::process_arguments
+
+  Inputs:
+    params - a list of function parameters
+    loc - location in the source
+    symbol_table - symbol table
+    init_code - code block, in which declaration of some arguments
+                may be added
+
+  Output: a list of expressions
+
+ Purpose: calls string_refine_preprocesst::process_operands with a
+          list of parameters.
+
+\*******************************************************************/
+
+exprt::operandst java_string_library_preprocesst::process_parameters(
+  const code_typet::parameterst &params,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table,
+  code_blockt &init_code)
+{
+  exprt::operandst ops;
+  for(const auto &p : params)
+  {
+    symbol_exprt sym(p.get_identifier(), p.type());
+    ops.push_back(sym);
+  }
+  return process_operands(ops, loc, symbol_table, init_code);
+}
+
+/*******************************************************************\
+
+Function: java_string_library_preprocesst::process_single_operand
+
+  Inputs:
+    processed_ops - the list of processed operands to populate
+    op_to_process - a list of expressions
+    loc - location in the source
+    symbol_table - symbol table
+    init_code - code block, in which declaration of some arguments
+                may be added
+
+ Purpose: Creates a string_exprt from the input exprt and adds it to
+          processed_ops
+
+\*******************************************************************/
+
+void java_string_library_preprocesst::process_single_operand(
+  exprt::operandst &processed_ops,
+  const exprt &op_to_process,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table,
+  code_blockt &init_code)
+{
+  member_exprt length(
+    op_to_process, "length", string_length_type(symbol_table));
+  member_exprt data(op_to_process, "data", string_data_type(symbol_table));
+  dereference_exprt deref_data(data, data.type().subtype());
+  string_exprt string_expr=fresh_string_expr(loc, symbol_table, init_code);
+  exprt string_expr_sym=fresh_string_expr_symbol(loc, symbol_table, init_code);
+#if 0
+  init_code.copy_to_operands(code_declt(string_expr_sym));
```

Comment:
What if we don't know??? It was never really clear whether this variable needed to be declared or not. Adding the declaration does not seem to change the results of our tests. Could you shed some light on this?